### PR TITLE
Cluster-autoscaler: add NodeReadyPredicate and allow unready nodes in CA

### PR DESCRIPTION
**Pull request by mwielgus**
Ref: #2228 

cc: @jszczepkowski @fgrzadkowski

### CODE REVIEW

File: cluster-autoscaler/scale_down_test.go [hunk 1]
```diff
@@ -275,6 +278,89 @@ func TestNoScaleDown(t *testing.T) {
...
+	n2 := BuildTestNode("n2", 1000, 1000)
+	SetNodeReadyState(n2, false, time.Time{})
+
+	p1 := BuildTestPod("p1", 100, 0)
+	p1.Annotations = map[string]string{
+		"kubernetes.io/created-by": RefJSON(&job),
+	}
+
+	p2 := BuildTestPod("p2", 800, 0)
+	p1.Spec.NodeName = "n1"
```
Comment (jszczepkowski) [inline cluster-autoscaler/scale_down_test.go]: nit: move 2 lines up

File: cluster-autoscaler/scale_down_test.go [hunk 2]
```diff
@@ -275,6 +278,89 @@ func TestNoScaleDown(t *testing.T) {
...
+	})
+	fakeClient.Fake.AddReactor("update", "nodes", func(action core.Action) (bool, runtime.Object, error) {
+		update := action.(core.UpdateAction)
+		obj := update.GetObject().(*apiv1.Node)
+		updatedNodes <- obj.Name
+		return true, obj, nil
+	})
+
+	provider := testprovider.NewTestCloudProvider(nil, func(nodeGroup string, node string) error {
+		deletedNodes <- node
```
Comment (jszczepkowski) [inline cluster-autoscaler/scale_down_test.go]: Where those channels get verified?

File: cluster-autoscaler/scale_down_test.go [hunk 3]
```diff
@@ -275,6 +278,89 @@ func TestNoScaleDown(t *testing.T) {
 	assert.Equal(t, n1.Name, getStringFromChan(deletedNodes))
 }
 
+func TestScaleDownNoMove(t *testing.T) {
+	deletedPods := make(chan string, 10)
```
Comment (jszczepkowski) [inline cluster-autoscaler/scale_down_test.go]: Where those pipes are verified?

Comment (mwielgus) [inline cluster-autoscaler/scale_down_test.go]: they are unused, removed.

File: cluster-autoscaler/scale_down_test.go (hunk 2 above)
Comment (mwielgus) [inline cluster-autoscaler/scale_down_test.go]: they are unused, removed.

File: cluster-autoscaler/scale_down_test.go [hunk 4]
```diff
@@ -275,6 +278,82 @@ func TestNoScaleDown(t *testing.T) {
...
+		switch getAction.GetName() {
+		case n1.Name:
+			return true, n1, nil
+		case n2.Name:
+			return true, n2, nil
+		}
+		return true, nil, fmt.Errorf("Wrong node: %v", getAction.GetName())
+	})
+	fakeClient.Fake.AddReactor("delete", "pods", func(action core.Action) (bool, runtime.Object, error) {
+		panic(fmt.Errorf("no delete is expected"))
```
Comment (jszczepkowski) [inline cluster-autoscaler/scale_down_test.go]: nit: wouldn't t.FailNow() be more appropriate?

File: cluster-autoscaler/cluster_autoscaler.go [hunk 5]
```diff
@@ -316,12 +315,6 @@ func run(_ <-chan struct{}) {
 					continue
 				}
 
-				// TODO: remove once all of the unready node handling elements are in place.
-				if err := CheckGroupsAndNodes(readyNodes, autoscalingContext.CloudProvider); err != nil {
```
Comment (jszczepkowski) [inline cluster-autoscaler/cluster_autoscaler.go]: Isn't method CheckGroupsAndNodes unused? If yes, can you remove it?

File: cluster-autoscaler/clusterstate/clusterstate.go [hunk 6]
```diff
@@ -401,15 +401,39 @@ func (csr *ClusterStateRegistry) GetUnregisteredNodes() []UnregisteredNode {
 
 // GetReadinessState gets readiness state for the node
 func GetReadinessState(node *apiv1.Node) (isNodeReady bool, lastTransitionTime time.Time, err error) {
-	for _, condition := range node.Status.Conditions {
-		if condition.Type == apiv1.NodeReady {
-			if condition.Status == apiv1.ConditionTrue {
-				return true, condition.LastTransitionTime.Time, nil
+	canNodeBeReady, readyFound := true, false
```
Comment (jszczepkowski) [inline cluster-autoscaler/clusterstate/clusterstate.go]: Please merge GetReadinessState and IsNodeReadyAndSchedulable into one method as they are very similar.

File: cluster-autoscaler/cluster_autoscaler.go (hunk 5 above)
Comment (mwielgus) [inline cluster-autoscaler/cluster_autoscaler.go]: Done.

File: cluster-autoscaler/scale_down_test.go (hunk 4 above)
Comment (mwielgus) [inline cluster-autoscaler/scale_down_test.go]: Done.

File: cluster-autoscaler/clusterstate/clusterstate.go (hunk 6 above)
Comment (mwielgus) [inline cluster-autoscaler/clusterstate/clusterstate.go]: Done.

**Comment by jszczepkowski** [comment]
LGTM

**Pull request merged**